MySQL: parser v2

QianKaiLin opened the pull request:
Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)

Link to ticket: https://github.com/OISF/suricata-verify/pull/2067

Describe changes:

- fix mysql parser bug.

- fix mysql logger bug.

- add mysql detection keywords `mysql.command` and `mysql.rows`

- add SV tests.

SV_REPO=
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2067

catenacyber reviewed (commented):
Thanks for the work so far, still more to do

CI : 🔴 could you fix it ?
Commits segmentation : I think we will squash all in the end, especially the commits like ` mysql: fix`, so we do not have a buggy point in our git history if we can avoid it
Commit messages : ok, we can see after squashing
Git ID set : ok
CLA : ❓ I do not have access
Doc update : 🟠 I think there is more to do, like a note in upgrade guide cc @jufajardini 
Redmine ticket : ok
Rustfmt : ok
Tests : thanks for the SV test
Dependencies added: none
Code : will look into it

catenacyber reviewed (changes requested):
CI is to be fixed

catenacyber commented on etc/schema.json:
```diff
@@ -2383,6 +2383,31 @@
+            "optional": true,
+            "properties": {
+                "version": {
+                    "type": "string"
```
Could you add `description` for these fields ?

catenacyber commented on suricata.yaml.in:
```diff
@@ -916,6 +917,12 @@ app-layer:
+    mysql:
+      enabled: no
+      # Stream reassembly size for MySQL. By default, track it completely.
+      stream-depth: 0
```
Why that default ?

QianKaiLin replied:
I copied from pgsql configuration, now i know what it means, i will remove it.

catenacyber commented on src/detect-mysql-rows.c:
```diff
@@ -0,0 +1,88 @@
+    return buffer;
+}
+
+void DetectMysqlRowsRegister(void)
```
You should be able to do these keywords in rust only ?
You can look at sip for a latest translation from C to rust for example ;-)

catenacyber commented on rust/src/mysql/logger.rs:
```diff
@@ -0,0 +1,69 @@
+    );
+    let result = log_mysql(tx_mysql, js);
+    if let Err(ref err) = result {
+        SCLogError!("----------- MySQL rs_mysql_logger failed. err is {:?}", err);
```
I am not sure we would want such logs in the end

catenacyber commented on rust/src/mysql/logger.rs:
```diff
@@ -0,0 +1,69 @@
+
+fn log_mysql(tx: &MysqlTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("mysql")?;
+    if let Some(version) = &tx.version {
```
Seems strange that all fields are optional... Is it really the case ?

catenacyber commented on etc/schema.json:
```diff
@@ -2383,6 +2383,31 @@
+                    "type": "array",
+                    "optional": true,
+                    "items": {
+                        "type": "string"
```
Could we have min_items so that we make sure we do not log an empty array ?

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+/// Detect
+/// Get the mysql query
+#[no_mangle]
+pub unsafe extern "C" fn SCMysqlTxGetCommand(
```
This should be moved to a detect.rs file

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+
+    if stream_slice.is_gap() {
+        state_safe.on_response_gap(stream_slice.gap_size());
+    } else if !stream_slice.is_empty() {
```
No need to recheck `!stream_slice.is_empty()`

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+            Ok(_) => return ALPROTO_MYSQL,
+            Err(nom7::Err::Incomplete(_)) => return ALPROTO_UNKNOWN,
+            Err(err) => {
+                SCLogError!("failed to probe response {:?}", err);
```
We should not this at error level, rather debug

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+}
+
+/// Probe for a valid mysql message
+pub fn probe(i: &[u8]) -> IResult<&[u8], ()> {
```
Does it need to be pub ?

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+                        if tx_completed {
+                            tx.complete = true;
+                        }
+                    }
```
Should we not create a new transaction with the response if ` self.find_or_create_tx() ` returns None ?

QianKaiLin replied:
Sorry, I don't understand the question.

catenacyber commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+    client_flags: u32,
+    version: Option<String>,
+    tls: bool,
+    command_code: Option<u8>,
```
(rather part of a transaction ? )

catenacyber commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+#[allow(dead_code)]
+pub const CLIENT_REMEMBER_OPTIONS: u32 = BIT_U32!(31);
+
+#[allow(dead_code)]
```
This seems used, why do we need `allow(dead_code)]` ?

catenacyber commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+impl From<u8> for FieldType {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => Self::Decimal,
```
Would device StringEnum8 help here ?

catenacyber commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+impl<'a> Drop for MysqlPacket<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            std::mem::drop(Box::from_raw(self.payload as *mut [u8]));
```
Why not use a Vec<u8> for ownership if we must ?

catenacyber commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+
+pub const SHA2_PASSWORD_REQUEST_PUBLIC_KEY: u8 = 2;
+pub const SHA2_PASSWORD_FAST_AUTH_SUCCESS: u8 = 3;
+pub const SHA2_PASSWORD_PERFORM_FULL_AUTH: u8 = 3;
```
Should it be 4 ? an enum rather than a list of constants ?

catenacyber commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+    let mut payload_len: usize = 0;
+    let mut rem = i;
+    let mut pkt_num = None;
+    loop {
```
Why do we loop to parse one header ?

catenacyber commented:
There is an amazing amount of work in here :-)

jufajardini commented:
I've adjusted the way the SV PR is passed, hopefully the step for preparing dependencies should work now.
Locally, I see that 3 of the sv tests are failing, expecting that the CI will highlight that, if it happens here, too.
For your next versions, you can follow this style. Paste the link to your SV PR in the `SV_BRANCH` option, like so:
```
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2067
```

:)

kudos on aaalll the work done so far, here!
For the CI failures, let us know if you need any guidance navigating and fixing them, ok?

jufajardini commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+        assert!(rem.is_empty());
+    }
+
+    // #[test]
+    //     let pkt: &[u8] = &[
+    //         22, 0, 0, 0, 23, 42, 2, 0, 0, 0, 1, 0, 0, 0, 0, 1, 8, 128, 16, 39, 0, 0, 0, 0, 0, 0, 5,
+    //         0, 0, 0, 25, 42, 2, 0, 0, 194, 0, 0, 0, 22, 83, 69, 76, 69, 67, 84, 32, 96, 117, 117,
+    //         105, 100, 96, 44, 96, 112, 97, 114, 101, 110, 116, 95, 105, 100, 96, 44, 96, 112, 97,
+    //         116, 104, 96, 44, 96, 110, 97, 109, 101, 96, 44, 96, 109, 101, 110, 117, 95, 108, 101,
+    //         118, 101, 108, 96, 44, 96, 104, 105, 100, 100, 101, 110, 96, 44, 96, 115, 111, 114,
```
Reminder to uncomment or remove these :P

jufajardini commented on rust/src/mysql/parser.rs:
```diff
@@ -0,0 +1,3447 @@
+
+    #[test]
+    fn test_parse_handshake_request() {
+        let pkt: &[u8] = &[
+            0x49, 0x00, 0x00, 0x00, 0x0a, 0x38, 0x2e, 0x34, 0x2e, 0x30, 0x00, 0x51, 0x00, 0x00,
+            0x00, 0x3e, 0x7d, 0x6a, 0x6a, 0x1a, 0x2d, 0x2b, 0x6b, 0x00, 0xff, 0xff, 0xff, 0x02,
+            0x00, 0xff, 0xdf, 0x15, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x61, 0x74, 0x54, 0x07, 0x62, 0x28, 0x5d, 0x21, 0x06, 0x44, 0x06, 0x62, 0x00, 0x63,
+            0x61, 0x63, 0x68, 0x69, 0x6e, 0x67, 0x5f, 0x73, 0x68, 0x61, 0x32, 0x5f, 0x70, 0x61,
+            0x73, 0x73, 0x77, 0x6f, 0x72, 0x64, 0x00,
```
As a good practice, when we have buffers of hex, especially long ones like that, could you add a comment indicating what is the buffer content?

jufajardini commented on rust/src/mysql/mysql.rs:
```diff
@@ -0,0 +1,1131 @@
+            self.tx_index_completed = index;
+        }
+        tx
+    }
```
nit: add new line

jufajardini reviewed (changes requested):
About the documentation:
- we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: https://github.com/OISF/suricata/commit/70ed9f91d84d)
- we should also document the log and keyword additions (one example: https://github.com/OISF/suricata/commit/70ed9f91d84d)

About the copyright note:
- all newly files should have it 
- If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?

As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!

victorjulien commented on etc/schema.json:
```diff
@@ -2388,23 +2388,27 @@
                 "command": {
-                    "type": "string"
+                    "type": "string",
+                    "description": "sql query statement or some utility commands like ping, quit etc."
```
nit: `quit, etc.`

victorjulien commented on rust/src/mysql/detect.rs:
```diff
@@ -0,0 +1,168 @@
+/* Copyright (C) 2022 Open Information Security Foundation
```
2024

victorjulien commented on rust/src/mysql/detect.rs:
```diff
@@ -0,0 +1,168 @@
+        if !rows.is_empty() {
+            let index = local_id as usize;
+            if let Some(row) = rows.get(index) {
+                if !row.is_empty() {
```
what does an empty row mean? Is it meaningful to match on it using `bsize:0;`?

victorjulien reviewed (changes requested):
a few comments, not a full review yet

QianKaiLin commented:
> About the documentation:
> 
>     * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> 
>     * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> 
> 
> About the copyright note:
> 
>     * all newly files should have it
> 
>     * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> 
> 
> As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!

I have added the document in upgrade and log and keyword

jufajardini commented:
> > About the documentation:
> > ```
> > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > 
> > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > About the copyright note:
> > ```
> > * all newly files should have it
> > 
> > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> 
> I have added the document in upgrade and log and keyword

Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)

QianKaiLin commented:
> > > About the documentation:
> > > ```
> > > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > 
> > > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > About the copyright note:
> > > ```
> > > * all newly files should have it
> > > 
> > > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> > 
> > 
> > I have added the document in upgrade and log and keyword
> 
> Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)

I have resolved conflicts, please check again, thx.

jufajardini commented:
> > > > About the documentation:
> > > > ```
> > > > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > > 
> > > > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > About the copyright note:
> > > > ```
> > > > * all newly files should have it
> > > > 
> > > > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> > > 
> > > 
> > > I have added the document in upgrade and log and keyword
> > 
> > 
> > Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)
> 
> I have resolved conflicts, please check again, thx.

Please, submit a new PR, to make it easier for us to review the cleaned up and rebased code, as well as to ensure that the CI checks will run against a rebased and updated SV PR :)
